internal/net/udp/packet_conn.go

```diff
@@ -0,0 +1,399 @@
+	return (&ListenConfig{}).Listen(network, laddr)
+}
+
+// readLoop dispatches packets to the proper connection, creating a new one if
+// necessary, until all connections are closed.
+func (l *listener) readLoop() {
+	defer l.readWG.Done()
+	defer close(l.readDoneCh)
+
+	buf := make([]byte, receiveMTU)
+
+	for {
+		n, raddr, err := l.pConn.ReadFrom(buf)
+		if err != nil {
+			l.errRead.Store(err)
+			return
+		}
+		conn, ok, err := l.getConn(raddr, buf[:n])
+		if err != nil {
+			continue
+		}
+		if ok {
+			_, _ = conn.buffer.WriteTo(buf[:n], raddr)
+		}
+	}
+}
+
+// getConn gets an existing connection or creates a new one.
+func (l *listener) getConn(raddr net.Addr, buf []byte) (*PacketConn, bool, error) {
+	l.connLock.Lock()
+	defer l.connLock.Unlock()
+	// If we have a custom resolver, use it.
+	if l.connResolver != nil {
+		if id, ok := l.connResolver(buf); ok {
+			if conn, ok := l.conns[id]; ok {
+				return conn, true, nil
+			}
+		}
+	}
+
+	// If we don't have a custom resolver, or we were unable to find an
+	// associated connection, fall back to remote address.
+	conn, ok := l.conns[raddr.String()]
+	if !ok {
+		if isAccepting, ok := l.accepting.Load().(bool); !isAccepting || !ok {
+			return nil, false, ErrClosedListener
+		}
+		if l.acceptFilter != nil {
+			if !l.acceptFilter(buf) {
+				return nil, false, nil
+			}
+		}
+		conn = l.newPacketConn(raddr)
+		select {
+		case l.acceptCh <- conn:
+			l.conns[raddr.String()] = conn
+		default:
+			return nil, false, ErrListenQueueExceeded
+		}
+	}
+	return conn, true, nil
+}
+
+// PacketConn is a net.PacketConn implementation that is able to dictate its
+// routing ID via an alternate identifier from its remote address. Internal
+// buffering is performed for reads, and writes are passed through to the
+// underlying net.PacketConn.
+type PacketConn struct {
+	listener *listener
+
+	raddr   net.Addr
+	rmraddr atomic.Bool
+	id      atomic.Value
+
+	buffer *idtlsnet.PacketBuffer
+
+	doneCh   chan struct{}
+	doneOnce sync.Once
+
+	writeDeadline *deadline.Deadline
+}
+
+// newPacketConn constructs a new PacketConn.
+func (l *listener) newPacketConn(raddr net.Addr) *PacketConn {
+	return &PacketConn{
+		listener:      l,
+		raddr:         raddr,
+		buffer:        idtlsnet.NewPacketBuffer(),
+		doneCh:        make(chan struct{}),
+		writeDeadline: deadline.New(),
+	}
+}
+
+// ReadFrom reads a single packet payload and its associated remote address from
+// the underlying buffer.
+func (c *PacketConn) ReadFrom(p []byte) (int, net.Addr, error) {
+	return c.buffer.ReadFrom(p)
+}
+
+// WriteTo writes len(p) bytes from p to the specified address.
+func (c *PacketConn) WriteTo(p []byte, addr net.Addr) (n int, err error) {
+	// If we have a connection identifier, check to see if the outgoing packet
+	// sets it.
+	if c.listener.connIdentifier != nil {
+		id := c.id.Load()
+		candidate, ok := c.listener.connIdentifier(p)
+		// If this is a new identifier, add entry to connection map.
```

Comment:
Yes, originally I checked to see if `id == nil`, but opted to not impose that restriction at this level given that it is a DTLS detail rather than a UDP one. We do pay a cost of calling the `connIdentifier()` though.

Reply:
yeah it seems like we should then make sure we only set it once though to avoid that cost. what do you think?

Reply:
That seems reasonable to me -- if we need to make it more generic to accommodate other use cases in the future we always can. I'll leave a comment reflecting that 👍🏻